examples/simpleplayer.cpp

```diff
@@ -163,16 +489,123 @@ class SimplePlayer
+        const VARectangle* srcRect,
+        const VARectangle* dstRect)
+    {
+        VAStatus vaStatus;
+        struct wl_buffer* buffer;
+        struct wl_callback* callback;
+        struct WaylanDisplay* const d = &m_waylandDisplay;
+        struct wl_callback_listener frame_callback_listener = { frameRedrawCallback };
+
+        if (m_redrawPending) {
+            wl_display_flush(d->display.get());
+            while (m_redrawPending) {
+                wl_display_dispatch(d->display.get());
+            }
+        }
+
+        if (!ensureWindow(dstRect->width, dstRect->height))
+            return false;
+
+        vaStatus = vaGetSurfaceBufferWl(m_vaDisplay, surface, VA_FRAME_PICTURE, &buffer);
+        if (vaStatus != VA_STATUS_SUCCESS)
+            return false;
+
+        wl_surface_attach(d->surface.get(), buffer, 0, 0);
+        wl_surface_damage(d->surface.get(), dstRect->x,
+            dstRect->y, dstRect->width, dstRect->height);
+        wl_display_flush(d->display.get());
+        m_redrawPending = true;
+        callback = wl_surface_frame(d->surface.get());
+        wl_callback_add_listener(callback, &frame_callback_listener, &m_redrawPending);
+        wl_surface_commit(d->surface.get());
+        return true;
+    }
+
+    bool waylandOutput(const SharedPtr<VideoFrame>& frame)
+    {
+        VARectangle srcRect, dstRect;
+        if (!ensureWindow(frame->crop.width, frame->crop.height))
+            return false;
+
+        srcRect.x = 0;
+        srcRect.y = 0;
+        srcRect.width = frame->crop.width;
+        srcRect.height = frame->crop.height;
+
+        dstRect.x = frame->crop.x;
+        dstRect.y = frame->crop.y;
+        dstRect.width = frame->crop.width;
+        dstRect.height = frame->crop.height;
+        return vaPutSurfaceWayland((VASurfaceID)frame->surface, &srcRect, &dstRect);
+    }
+
+    bool ensureWindow(unsigned int width, unsigned int height)
+    {
+        struct WaylanDisplay* const d = &m_waylandDisplay;
+
+        if (!d->surface) {
+            d->surface.reset(wl_compositor_create_surface(d->compositor.get()), wl_surface_destroy);
+            if (!d->surface)
+                return false;
+        }
+
+        if (!d->shell_surface) {
+            d->shell_surface.reset(wl_shell_get_shell_surface(d->shell.get(), d->surface.get()),
+                wl_shell_surface_destroy);
+            if (!d->shell_surface)
+                return false;
+            wl_shell_surface_set_toplevel(d->shell_surface.get());
+        }
+        return true;
+    }
+#endif
+
     SharedPtr<NativeDisplay> m_nativeDisplay;
     VADisplay m_vaDisplay;
-    Window   m_window;
     SharedPtr<IVideoDecoder> m_decoder;
     SharedPtr<DecodeInput> m_input;
     int m_width, m_height;
+    bool m_eos;
+    int m_drmFd;
+    std::ofstream m_ofs;
+    uint64_t m_frameNum;
+    SimplePlayerParameter m_parameters;
+#ifdef __ENABLE_X11__
+    SharedPtr<Display> m_display;
```

Comment:
could we have some abstract and isolate all X11/Wayland specific operation in class?

Reply:
this code is used in fast-boot, so we try to keep the code as small as possible, and not to define classes.